src/feature/relay/dns.c

```diff
@@ -2101,6 +2101,35 @@ dump_dns_mem_usage(int severity)
       (unsigned)hash_mem);
 }
 
+/* Do a round of OOM cleanup on all DNS entries. Return the amount of removed
+ * bytes. It is possible that the returned value is lower than min_remove_bytes
+ * if the caches get emptied out so the caller should be aware of this. */
+size_t
+dns_cache_handle_oom(time_t now, size_t min_remove_bytes)
+{
+  time_t time_inc = 0;
+  size_t bytes_removed = 0;
+  size_t current_size = dns_cache_total_allocation();
+
+  do {
+    /* If no DNS entries left, break loop. */
+    if (!dns_cache_entry_count())
+      break;
+
+    /* Get cutoff interval and remove entries. */
+    time_t cutoff = now + time_inc;
+    purge_expired_resolves(cutoff);
+
+    /* Update amount of bytes removed and array size. */
+    bytes_removed += current_size - dns_cache_total_allocation();
+    current_size -= bytes_removed;
```

Comment:
I think there is a problem here. Since `bytes_removed` increment every loop, is it possible you'll get offset as you go like this (considering initial state to be `bytes_removed = 0` and `current_size = 100` and every loop, let say we remove 1 entry of size `10`:

```
bytes_removed - current_size
0 - 100
10 - 90
20 - 70
30 - 40
40 - 0
```